### DISCUSSION TITLE
How to reflect the `tuple` types to GraphQL?

### DISCUSSION THREAD

**Discussion by vpetrovykh** [geldata/gel#3154] [Ideas]
Re: issue #1252

Tuples can be unnamed, where the elements are more like indices or named, where elements are more like properties of a FreeObject. However, because tuples can contain any combination of data types, we can't reflect them as `List` even if they are unnamed.

### Read as JSON

One option is to reflect them as JSON scalars, that way we don't need to worry about the internal structure as we'll just get the entire tuple value as a single blob. In practice, it's very likely that treating the tuple as a single piece of data is what is typically
intended. Otherwise it's quite likely that an object would have been used instead in the first place.

Queries are simple: `query {Foo {val}}`.

### Read as an Object

Another option is that we actually reflect all tuples into some `Object` and simply use indices as field names for unnamed tuples.
This would give the benefit of exposing the structure of the tuple to the GraphQL schema and enable a user to see it. The downside is that now we have a GraphQL object type like `MyTup` and when we query it we must specify which fields we want. We can also rename the fields. So we now must support something like this:

```
query {
 Foo {
   val_tuple {
     a
     b
     result: c
   }
 }
}
```

Compiled as something like:
```
select {
 multi Foo := (
   select Foo{
     val_tuple := (
       a := .val.a,
       b := .val.b,
       result := .val.c
     ),
   }
 )
}
```

As one can imagine, if there are nested structures, this becomes a bit more complicated:
- As long as tuples are nested in tuples, we would be forced to allow the user to pick and choose which fields to get and also what to call them. But generally it's not too bad and very similar to just nested shapes.
- If we have a nested array then we would still need to expose the underlying structure. This means that we'd be unpacking and aggregating nested arrays.

  Consider this schema (because the general case should handle all the features present in it):
  ```
  type Foo {
      val -> tuple<
          a: str, 
          b: bool, 
          c: array<tuple<x: int64, y: int64, z: int64>>, 
          d:int64>;
  }
  ```

  Now for a general GraphQL query that makes use of selecting part of the fields and renaming them:
  ```
  query {
   Foo {
     // need to give a name that is different from the
     // actual property name because we cannot change type
     // of the existing property by overwriting it
     val_tuple {
       a
       result: c_tuple {x y}
       }
     }
   }
  }
  ```

  Would have to be compiled into something like this:    
  ```
  select {
   multi Foo := (
     select Foo{
       val_tuple := (
         a := .val.a,
         result := array_agg((
           with _tup := array_unpack(.val.c)
           select (x := _tup.x, y := _tup.y)
         ))
       ),
     }
   )
  }
  ```

### Input as JSON

Now all the reasons that made JSON blob a simpler format to handle when _reading_ tuples makes it a bad format for inputting them. There's no internal structure declared in the schema, so all validation would have to be done in EdgeQL, that's checking for `null` in JSON, because `{}` are illegal as tuple elements and also just checking that the internal JSON structure actually conforms to the tuple shape. It is technically doable, but pretty messy and will need a ton of `assert_exists`. Among other things it would require raising exceptions in EdgeQL when the input is not in the right shape (which currently would need a custom internal function to be used as we don't have exceptions yet).

### Input as Object

This is great for input because it allows us to use GraphQL validation to make sure that the input conforms to the expected tuple shape. All the fields are present, they are of correct types, no `null`, etc.

## Recommendation

Given the options it appears that going with an asymmetrical approach would be best: JSON for queries, Objects for input data in mutations. In practice, the difference in usage is very marginal for a person, namely that JSON can only be submitted as a variable, not a literal in the mutation body, whereas the Object could be submitted in either way. The actual data submitted and data being received would be the same, though.

The JSON output is good enough for most tuple use-cases and avoids coding complexity for features that are fairly marginal
(cherry-picking tuple data). Plus we can always restructure the tuple for reading by adding a computed property if that's necessary.

The Object input is great because there's no extra complexity in implementing a whole validation layer in EdgeQL, making use of GraphQL validation instead.

Unfortunately we cannot simply add a union type of JSON and Object, because mixing scalars and objects is forbidden in GraphQL.

**Comment by msullivan** [comment]
Would it be possible to do "output as object", and then *require* that the user specifies all fields and does not rename them? 

That would give us consistency with very little initial implementation burden, and we can always fix it later.

Downside is that it does seem more annoying to use, though. 

(Can we expose both output mechanisms, "json" and "object but we error if you are a pain in the butt"?)

**Reply by vpetrovykh** [reply]
This kinda goes against the whole GraphQL a-la-carte philosophy. You're supposed to be able to specify which output fields you want.

**Comment by elprans** [comment]
I don't think "Input as JSON" is that hard.  We already have the machinery to do json -> tuple casts, including type and null checks.
